[Model 8] REST API polishing, cleanup and extensions

afein opened the pull request:
This PR is focused in 2 major changes to the model REST API:
- Model GetXMetric methods now receive specific Request structs as parameters.
- Added new Getters,  endpoints and tests for extracting all nodes, namespaces, pods and containers from the model. This should allow API consumers to seamlessly 

\cc @rjnagal @vmarmol

mvdan commented on api/v1/model_handlers.go:
```diff
@@ -35,12 +36,19 @@ func (a *Api) RegisterModel(container *restful.Container) {
 		Filter(compressionFilter).
-		Doc("Show which metrics are available").
-		Operation("clusterMetrics"))
+		Doc("Get a list of all available metrics for the Cluster entity.").
```
Period should probably go.

mvdan commented on api/v1/model_handlers.go:
```diff
@@ -50,8 +58,25 @@ func (a *Api) RegisterModel(container *restful.Container) {
+	ws.Route(ws.GET("/nodes/{node-name}").
+		To(a.availableMetrics).
+		Filter(compressionFilter).
+		Doc("Get a list of all available metrics for a Node entity.").
```
Same here.

mvdan commented on api/v1/model_handlers.go:
```diff
@@ -267,11 +388,11 @@ func (a *Api) freeContainerMetrics(request *restful.Request, response *restful.R
-func parseRequestStartParam(request *restful.Request, response *restful.Response) time.Time {
+// parseRequestParam parses a time.Time from a named QueryParam.
+// parseRequestParam receives a request and a response as inputs, and returns the parsed time.
+func parseRequestParam(param string, request *restful.Request, response *restful.Response) time.Time {
```
nit: shouldn't extra arguments be at the end? Also, looks like its doc needs updating.

afein replied:
I think that parseRequestParam("start", request, response) is more readable than parseRequestParam(request, response, "start"), especially for consecutive calls with the same request/response arguments.

mvdan commented on model/impl.go:
```diff
@@ -161,54 +160,136 @@ func (rc *realCluster) GetPodContainerMetric(namespace string, pod_name string,
+		return res
+	}
+
+	for key, _ := range podref.Containers {
```
nit: `, _` can be removed

afein commented:
updated the API to provide metrics under /metrics for each entity, thus enhancing browsability of the API and allowing a future /stats extension to be added to each entity.
Also addressed most of mvdan's comments.

mvdan commented:
retest this please

mvdan commented:
Jenkins is trying this on kubernetes 0.20.0, this is weird.
